api/pkg/transformer/feast/feature_retriever.go

```diff
@@ -206,95 +230,88 @@ func addSeries(entityName string, series []*types.Value, entities []feast.Row) [
+		featureTableSpec: featureTableSpec,
+		entityList:       entityList,
+		defaultValues:    fr.defaultValues,
 
-	batchResultChan := make(chan batchResult, numOfBatch)
-	columns := getColumnNames(featureTableSpec)
-	entityIndices := getEntityIndicesFromColumns(columns, featureTableSpec.Entities)
-	for i := 0; i < numOfBatch; i++ {
-		startIndex := i * fr.options.BatchSize
-		endIndex := len(entityNotInCache)
-		if endIndex > startIndex+fr.options.BatchSize {
-			endIndex = startIndex + fr.options.BatchSize
-		}
-		batchedEntities := entityNotInCache[startIndex:endIndex]
+		feastClient: fr.feastClient,
 
-		go func(project string, entityList []feast.Row, columns []string) {
-			feastRequest := feast.OnlineFeaturesRequest{
-				Project:  project,
-				Entities: entityList,
-				Features: features,
-			}
-			startTime := time.Now()
-			feastResponse, err := fr.feastClient.GetOnlineFeatures(ctx, &feastRequest)
-			durationMs := time.Now().Sub(startTime).Milliseconds()
-			if err != nil {
-				feastLatency.WithLabelValues("error").Observe(float64(durationMs))
-				feastError.Inc()
+		cache:        fr.cache,
+		cacheEnabled: fr.options.CacheEnabled,
+		cacheTTL:     fr.options.CacheTTL,
+		logger:       fr.logger,
 
-				batchResultChan <- batchResult{featuresData: nil, err: err}
-				return
-			}
-			feastLatency.WithLabelValues("success").Observe(float64(durationMs))
+		statusMonitoringEnabled: fr.options.StatusMonitoringEnabled,
+		valueMonitoringEnabled:  fr.options.ValueMonitoringEnabled,
+	}
+}
 
-			fr.logger.Debug("feast_response", zap.Any("feast_response", feastResponse.Rows()))
+func (fc *feastCall) do(
+	ctx context.Context,
+	features []string,
+	columns []string,
+	entityIndexMap map[int]int,
+) batchResult {
+	feastRequest := feast.OnlineFeaturesRequest{
+		Project:  fc.featureTableSpec.Project,
+		Entities: fc.entityList,
+		Features: features,
+	}
 
-			entityFeaturePairs, err := fr.buildFeastFeaturesData(ctx, feastResponse, featureTableSpec.Project, columns, entityIndices)
-			if err != nil {
-				batchResultChan <- batchResult{featuresData: nil, columnTypes: nil, err: err}
-				return
-			}
+	startTime := time.Now()
+	feastResponse, err := fc.feastClient.GetOnlineFeatures(ctx, &feastRequest)
+	durationMs := time.Now().Sub(startTime).Milliseconds()
+	if err != nil {
+		feastLatency.WithLabelValues("error").Observe(float64(durationMs))
+		feastError.Inc()
 
-			var featuresData transTypes.ValueRows
-			for _, data := range entityFeaturePairs {
-				featuresData = append(featuresData, data.value)
-			}
+		return batchResult{featuresData: nil, err: err}
+	}
+	feastLatency.WithLabelValues("success").Observe(float64(durationMs))
 
-			if fr.options.CacheEnabled {
-				if err := insertMultipleFeaturesToCache(fr.cache, entityFeaturePairs, project, fr.options.CacheTTL); err != nil {
-					fr.logger.Error("insert_to_cache", zap.Any("error", err))
-				}
-			}
+	fc.logger.Debug("feast_response", zap.Any("feast_response", feastResponse.Rows()))
```

Comment:
Can this be removed? even though it's debug `feastResponse.Rows()` can still be executed.